### ISSUE TITLE
Load EfficientNet Keras implementation with OpenCV dnn

### ISSUE THREAD

**Issue by d-cosin** [opencv/opencv#18722]
##### System information (version)

- OpenCV => 4.4.0.44 or 4.2
- Operating System / Platform => Ubuntu 18.04.5
- Compiler => Python 3

##### Detailed description

I am trying to load a frozen .pb model through `cv2.dnn.readNetFromTensorflow('frozen_model.pb')`. The model is the Keras implementation of EfficientNet, trained as shown [in this tutorial](https://keras.io/examples/vision/image_classification_efficientnet_fine_tuning/).

When I try to load the frozen .pb model into ``cv2.dnn.readNetFromTensorflow`, it throws the following exception:

```
error: OpenCV(4.4.0) /tmp/pip-req-build-6amqbhlx/opencv/modules/dnn/src/tensorflow/tf_importer.cpp:553: error: (-2:Unspecified error) Input layer not found: sequential/efficientnetb0/normalization/Reshape_1/ReadVariableOp/resource in function 'connect'
```

##### Steps to reproduce

1. Train a EfficientNet model in Keras as described [in this tutorial](https://keras.io/examples/vision/image_classification_efficientnet_fine_tuning/)
2. Convert the generated savedModel to a frozen graph .pb file following [this tutorial](https://leimao.github.io/blog/Save-Load-Inference-From-TF2-Frozen-Graph/). I uploaded the script [in this gist](https://gist.github.com/d-cosin/753748c488671bd80cd48f90683a79a3)
3. Run `cv2.dnn.readNetFromTensorflow('frozen_model.pb')`

**Comment by dkurt** [comment]
You have to freeze model before usage. "ReadVariableOp" means model in training mode with Variables instead Constants.

**Comment by d-cosin** [comment]
@dkurt Sorry, but could you please elaborate why the model is not frozen and in training mode? I ask because I used the same script to freeze a different neural network architecture implemented in Keras and, in this case, `cv2.dnn.readNetFromTensorflow`was able to load it.

**Comment by naranjuelo** [comment]
Hi @d-cosin, were you able to solve the problem? I'm facing the same issue

**Comment by d-cosin** [comment]
Hi @naranjuelo. I was not able to solve it.

**Comment by LupusSanctus** [comment]
@d-cosin, could you, please, attach your saved TF trained model (before it was frozen)?

**Comment by d-cosin** [comment]
@LupusSanctus sure. I am uploading a .zip, because GitHub does not support .pb files.

[frozen_graph.pb.zip](https://github.com/opencv/opencv/files/5695894/frozen_graph.pb.zip)

**Comment by LupusSanctus** [comment]
@d-cosin, thank you for the attached .pb! Judging by its graph, the initial problem with ``ReadVariableOp`` was solved, probably, with its transformation. However, there is another problem: ``
sequential/efficientnet-b0/stem_activation/IdentityN``.
I've reproduced the problems by the simple basic example of TF pretrained EfficientNetB0. Let's explore each below point:
1. ``tf_importer.cpp:553: error: (-2:Unspecified error) Input layer not found: efficientnetb0/normalization/Reshape_1/ReadVariableOp/resource in function 'connect'``
got the problem (the same as in issue) after the model was frozen with provided in the issue [script](https://gist.github.com/d-cosin/753748c488671bd80cd48f90683a79a3). The problem is not on OpenCV side. The frozen graph shouldn't contain operations as input.  
To avoid it, I've used TF ``transform_graph`` tool. To build it with bazel (this step may be skipped if it has been already built) execute:
```
bazel build tensorflow/tools/graph_transforms:transform_graph
```

After it was successfully built, run:
```
bazel-bin/tensorflow/tools/graph_transforms/transform_graph --in_graph=<your_path>/effnet_models/efficientNetB0_frozen.pb --out_graph=<your_path>/effnet_models/optimized_frozen_inference_graph.pb --inputs="x" --outputs="Identity" --transforms='remove_nodes(op=Identity) strip_unused_nodes fold_constants(ignore_errors=true) fold_batch_norms fold_old_batch_norms'
```
Now the graph looks better. But there are plenty of ``IdentityN`` nodes. Thus, after its loading with OpenCV DNN API (``cv2.dnn.readNetFromTensorflow``) this will cause the following problem:
2. ``Error: dnn.cpp:614: error: (-2:Unspecified error) Can't create layer "efficientnetb0/stem_activation/IdentityN" of type "IdentityN" in function 'getLayerInstance'``
here is something wrong with activation. Let's look into [TF EfficientNet implementation](https://github.com/tensorflow/tensorflow/blob/v2.4.0/tensorflow/python/keras/applications/efficientnet.py#L201):
```
def EfficientNet(
    ...,
    activation='swish',
    ...)
```
So, ``activation='swish'`` is [used](https://github.com/tensorflow/tensorflow/blob/v2.4.0/tensorflow/python/keras/applications/efficientnet.py#L331) in the model:
```
  x = layers.Activation(activation, name='stem_activation')(x)
```
as we can see, in [EfficientNetB0](https://github.com/tensorflow/tensorflow/blob/v2.4.0/tensorflow/python/keras/applications/efficientnet.py#L517-L540) no activation is set, so activation=['swish'](https://www.tensorflow.org/api_docs/python/tf/keras/activations/swish).

It turns out, there is a bunch of issues about strange TF EfficientNet behaviour - one of them is https://github.com/tensorflow/tensorflow/issues/43576 (finally it refers to https://github.com/tensorflow/tensorflow/issues/40166, here now is the main activity). And there is an interesting note: https://github.com/tensorflow/tensorflow/issues/43576#issuecomment-699251854. So, let's try to set ``EfficientNetB0`` activation to "relu":

```
EfficientNetB0(include_top=True, weights='imagenet', activation="relu", classes=1000)
```

Now, after repeating above described optimization -> transformation steps, we can finally successfully load the model with ``cv2.dnn.readNetFromTensorflow``.


Probable overall WA: you may also try to use https://github.com/qubvel/efficientnet.

**Comment by asmorkalov** [comment]
https://github.com/opencv/opencv/issues/19696

**Comment by advanpix** [comment]
It is a real pain to see that the issue gets low priority from developers (removed from 4.5.1, 4.5.2 milestones, etc.) 
EfficientNet is the current state-of-the-art family of architectures. Lots of applications are using it as a backbone - from detection to segmentation. Lack of its support grows to be a show stopper for using the OpenCV.

Definitely OpenCV would greatly benefit from its support. I might be wrong, but this boils down to implementing just another activation function, swish  = x*sigmoid(x) for CPU/GPU.

By any chance, do the developer provide some guide on how to add new activation function to dnn module?

**Comment by asmorkalov** [comment]
@advanpix swish activation is supported by OpenCV dnn module, even on NVIDIA GPUs. The issue is related to TensorFlow model parser that handles the layer incorrectly. As work around you can try ReLU as recommened above or try to convert model to ONNX format. ONNX parser supports swish activation and should work as expected.

**Comment by teaglin** [comment]
> @advanpix swish activation is supported by OpenCV dnn module, even on NVIDIA GPUs. The issue is related to TensorFlow model parser that handles the layer incorrectly. As work around you can try ReLU as recommened above or try to convert model to ONNX format. ONNX parser supports swish activation and should work as expected.

I tried that here. https://github.com/opencv/opencv/issues/19696 Removed swish and used RELU, but getting other issues.
